### PR TITLE
IGNITE-17889 SQL Calcite: Avoid full index scans in case of NULLs in search bounds

### PR DESCRIPTION
**Pull request by alex-plekhanov**
Thank you for submitting the pull request to the Apache Ignite.

In order to streamline the review of the contribution 
we ask you to ensure the following steps have been taken:

### The Contribution Checklist
- [ ] There is a single JIRA ticket related to the pull request. 
- [ ] The web-link to the pull request is attached to the JIRA ticket.
- [ ] The JIRA ticket has the _Patch Available_ state.
- [ ] The pull request body describes changes that have been made. 
The description explains _WHAT_ and _WHY_ was made instead of _HOW_.
- [ ] The pull request title is treated as the final commit message. 
The following pattern must be used: `IGNITE-XXXX Change summary` where `XXXX` - number of JIRA issue.
- [ ] A reviewer has been mentioned through the JIRA comments 
(see [the Maintainers list](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute#HowtoContribute-ReviewProcessandMaintainers)) 
- [ ] The pull request has been checked by the Teamcity Bot and 
the `green visa` attached to the JIRA ticket (see [TC.Bot: Check PR](https://mtcga.gridgain.com/prs.html))

### Notes
- [How to Contribute](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute)
- [Coding abbreviation rules](https://cwiki.apache.org/confluence/display/IGNITE/Abbreviation+Rules)
- [Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Coding+Guidelines)
- [Apache Ignite Teamcity Bot](https://cwiki.apache.org/confluence/display/IGNITE/Apache+Ignite+Teamcity+Bot)

If you need any help, please email dev@ignite.apache.org or ask anу advice on http://asf.slack.com _#ignite_ channel.

### CODE REVIEW

**Review by ivandasch — CHANGES_REQUESTED** [review]
Looks good in general.
Could you please address some minor issues?

File: modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/exec/exp/IgniteSqlFunctions.java [hunk 1]
```diff
@@ -154,6 +154,27 @@ public static String toString(ByteString b) {
...
+    public static Object greatest2(Object arg0, Object arg1) {
+        return leastOrGreatest(false, arg0, arg1);
+    }
+
+    /** */
+    private static Object leastOrGreatest(boolean least, Object arg0, Object arg1) {
+        if (arg0 == null || arg1 == null)
+            return null;
+
+        if (((Comparable<Object>)arg0).compareTo(arg1) < 0)
```
[the remark this answers is not in the document]
Comment (alex-plekhanov) [reply modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/exec/exp/IgniteSqlFunctions.java:172]: These methods only used to build index search bounds. All indexed values should be comparable.

File: modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/exec/exp/ExpressionFactoryImpl.java [hunk 2]
```diff
@@ -765,16 +797,20 @@ public RangeIterableImpl(List<RangeCondition<Row>> ranges, Comparator<Row> compa
...
         }
 
         /** {@inheritDoc} */
         @Override public Iterator<RangeCondition<Row>> iterator() {
             ranges.forEach(b -> ((RangeConditionImpl)b).clearCache());
 
-            if (ranges.size() == 1)
-                return ranges.iterator();
+            if (ranges.size() == 1) {
+                if (((RangeConditionImpl)ranges.get(0)).skip())
```
[the remark this answers is not in the document]
Comment (alex-plekhanov) [reply modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/exec/exp/ExpressionFactoryImpl.java:809]: `skip` is a details of internal implementation, not related somehow to the abstract range conditions, and doesn't required outside of `ExpressionFactoryImpl`.

[the remark this answers is not in the document]
Comment (ivandasch) [reply modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/exec/exp/ExpressionFactoryImpl.java:809]: Ok, make sense. But interface is also our own. Casting looks a little bit ugly.

File: modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/exec/exp/IgniteSqlFunctions.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (ivandasch) [reply modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/exec/exp/IgniteSqlFunctions.java:172]: Let's make and assertion here?

**Review by ivandasch — APPROVED** [review]

File: modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/exec/exp/IgniteSqlFunctions.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (alex-plekhanov) [reply modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/exec/exp/IgniteSqlFunctions.java:175]: Assert added

**Pull request closed without merging**